### PR TITLE
TST: Backend switching API (don't merge)

### PR DESCRIPTION
**Pull request by timhoffm**
## PR Summary

This PR creates a possible API for backend switching. The API is laid out in the [last commit](https://github.com/matplotlib/matplotlib/pull/11612/commits/eec27b040847b73cc782d5425fc9767ea9e66e03). Previous commits are from https://github.com/matplotlib/matplotlib/pull/11600/ and only contain the actual switching logic.

The code is not fully tested. It's main purpose is to give an idea how the following API can be implemented:

- There is a new `rcParams['default_backends']` that accepts a list of possible backends to try out.
- The current behavior for setting the default backend via `rcParams['backend']` is deprecated.
- Reading `rcParams['backend'] is deprecated. Users should use `rcParams['default_backends']` or `get_backend()` depeding on their needs.
- `matplotlib.use()` is deprecated in favor of `set_backend()` -- this is independent of the other API changes but adds to consistency.

### Setting the default backend
new API:
~~~
rcParams['default_backends'] = ['qt5agg', 'qt4agg']
~~~
current API (deprecated):
~~~
rcParams['backend'] = 'qt5agg'  # only one supported
WARNING: The rcParam 'backend' is deprecated. Use 'default_backends' instead.
~~~

### Setting the current backend
new API:
~~~
matplotlib.set_backend('qt5agg')
~~~
current API (deprecated):
~~~
matplotlib.use('qt5agg')
WARNING: matplotlib.use is deprecated. Use matplotlib.set_backend instead.
~~~

### Querying the current backend:
new API:
~~~
matplotlib.get_backend()
~~~
current API (deprecated):
~~~
backend = matplotlib.rcParams['backend']  # returns the current or default backend
WARNING: The rcParam 'backend' is deprecated. Use the rcParam 'default_backends' to access the defaults. Use get_backend() to get the current backend."
~~~

### CODE REVIEW

**Comment by tacaswell** [comment]
The `matplotlib.use` API has been there from 2004 (goes all the way back to e34a333d00814124d3e19d462b9d78ac35e7a49a) and it is in a lot of code / documentation / examples.  I am not convinced that the slight improvement in the verbiage is worth disrupting a considerable fraction of our users.  While if we were going to be starting from a clean slate I agree `set_backend` is a better name, I am :-1: on deprecating `use`.  

I am 50/50 on adding `set_backend` as an alias for `use`, it is a better more guessable name, but now we have 2 versions of the call...

Similarly, storing the current backend in `rcParams` has been there as long, ex the current implementation of `get_backend` is:

https://github.com/matplotlib/matplotlib/blob/e3042ab896ff59b559f13a55a7c08b8ce616b120/lib/matplotlib/__init__.py#L1380-L1382


I think I am on board with pushing people to `get_backend` instead of `rcParams['backend']`, but that is going to need a long (multi-year) deprecation cycle. 

I am :+1: on adding `'default_backends'` as a parameter, but (as discussed on the call today) we may want to move things like backend, interactive, etc out of rcParams all together so we should have a plan for that before we put it in.

File: lib/matplotlib/pyplot.py [hunk 1]
```diff
@@ -2401,7 +2391,11 @@ def _autogen_docstring(base):
 install_repl_displayhook()
 
 # Set up the backend.
-switch_backend(rcParams["backend"])
+for candidate in rcParams['default_backends']:
+    try:
+        switch_backend(candidate)
+    except ImportError:
+        continue
```
Comment (anntzer) [inline lib/matplotlib/pyplot.py]: probably need to do something if no backend is successfully imported.

File: lib/matplotlib/__init__.py [hunk 2]
```diff
@@ -909,8 +911,17 @@ def __setitem__(self, key, val):
...
+    def _set_current_backend(self, backend):
+        dict.__setitem__(self, 'backend', backend)
+
     def __getitem__(self, key):
-        if key in _deprecated_map:
+        if key == 'backend':
+            cbook.warn_deprecated('3.0',
+                "The rcParam 'backend' is deprecated. Use the rcParam "
+                "'default_backends' to access the defaults. Use get_backend() "
+                "to get the current backend.")
```
Comment (anntzer) [inline lib/matplotlib/__init__.py]: But then you need to store the current backend somewhere else and update the implementation of get_backend accordingly.

**Comment by jklymak** [comment]
@timhoffm I agree with your suggestions above.  OTOH for now we want to get backend switching in, and the existing rcParam stuffing, while suboptimal, is the existing API.  So for 3.0 I think we decided to move along w/ @anntzer PR (plus or minus any review changes) and then we can re-open the deprecation/API refinement for 3.1 (due December-ish).

**Comment by timhoffm** [comment]
Superseeded by #11600

**Pull request closed without merging**